Fixes Issue2110: Set transaction `type` at `output` level

hewison-chris opened the pull request:
To enable a spendable refund output in a freeze transaction we set the `Payment`, `Freeze` or `Coinbase` enum on each output.

hewison-chris committed:
Set tx `type` for each tx `output`

To enable a `Payment` refund output in a freeze transaction we set the
`Payment`, `Freeze` or `Coinbase` enum on each output.

MichaelKim20 commented on source/agora/consensus/data/Transaction.d:
```diff
@@ -169,6 +166,14 @@ unittest
+/// Indicates if output is frozen (staking), payment (spending) or coinbase (rewards / fees)
+public enum OutputType : ubyte
+{
+    Payment,
+    Freeze,
+    Coinbase,
```
Can we find the refund output?

hewison-chris replied:
When creating a transaction with one or more `freeze` outputs, the refund output will be a single `output` of type `Payment` if needed (within the same transaction).

Geod24 commented on source/agora/utils/TxBuilder.d:
```diff
@@ -257,7 +257,7 @@ public struct TxBuilder
     ***************************************************************************/
 
-    public Transaction sign (in TxType type = TxType.Payment, const(ubyte)[] data = [],
+    public Transaction sign (in OutputType type = OutputType.Payment, const(ubyte)[] data = [],
```
This doesn't make sense anymore. We need to re-think the API.

hewison-chris replied:
It felt a bit strange adding the OutputType to the API but works out quite well in practice.  Maybe the name of the param should be better. How about `primary_output_type`?

hewison-chris replied:
Changed to `outputs_type` and added more details in the param documentation of how it is used.

hewison-chris commented:
The last commit is to enable the `Integration` test to run as an updated Faucet is required. When green and approved for merging I will drop it and once merged we will need to update Faucet PR: https://github.com/bosagora/faucet/pull/87.

hewison-chris commented on source/agora/consensus/data/Transaction.d:
```diff
@@ -159,6 +136,29 @@ public struct Transaction
+
+    pure nothrow @nogc:
+
+    /// A `Freeze` transaction is one that has one or more `Freeze` outputs
+    /// If there is more than one output then it is allowed to have a single
+    ///  `Payment` output for a refund of any amount
+    public bool isFreeze () const
+    {
+        return this.outputs.any!(o => o.type == OutputType.Freeze);
+    }
+
+    /// A `Coinbase` transaction is one that has one or more `Coinbase` outputs
+    /// However if all outputs are not `Coinbase` then it will fail validation
+    public bool isCoinbase () const
+    {
+        return this.outputs.any!(o => o.type == OutputType.Coinbase);
+    }
+
+    /// A `Payment` transaction is one that has outputs of type `Payment`
+    public bool isPayment () const
+    {
+        return this.outputs.all!(o => o.type == OutputType.Payment);
+    }
```
@MichaelKim20  I added some more documentation here about the Transaction and output type relationship.

linked0 commented on doc/Database.md:
```diff
@@ -11,14 +11,14 @@ For more informations, see [this description](https://github.com/bosagora/agora/
-| `amount`        | INTEGER  | Amount    | NOT NULL    | See `Output` struct                                    |
-| `locktype`      | INTEGER  | LockType  | NOT NULL    | See `Lock` struct                                      |
-| `lock`          | BLOB     | ubyte[]   | NOT NULL    | See `Lock` struct                                      |
+| Field name      | SQL Type | D type      | Attributes  | Comment                                                |
```
The lines of the table are not aligned.

hewison-chris replied:
Updated

hewison-chris commented on source/agora/consensus/validation/Transaction.d:
```diff
@@ -129,8 +122,11 @@ public string isInvalidReason (
-    if (tx.type == TxType.Freeze)
+    if (tx.isFreeze)
     {
+        if (tx.outputs.count!(o => o.type == OutputType.Payment) > 1)
+            return "Transaction: Freeze cannot have multiple refund payment outputs";
+
```
@MichaelKim20 Here is validation for more than single `Payment` output in a `Freeze` transaction.

linked0 commented on source/agora/consensus/data/Transaction.d:
```diff
@@ -159,6 +136,29 @@ public struct Transaction
+    /// A `Freeze` transaction is one that has one or more `Freeze` outputs
+    /// If there is more than one output then it is allowed to have a single
+    ///  `Payment` output for a refund of any amount
+    public bool isFreeze () const
```
The name is kind of confusing which makes me think that all the outputs are `Freeze`. `hasFreeze`, `containsFreeze` would be better.

hewison-chris replied:
The transaction is still having the concept of `freeze` though. Although there may be a `Payment` output included for refund, these helpers are used where the logic is at the transaction level rather than the outputs within.

linked0 replied:
Got it. Thanks.

linked0 commented on source/agora/consensus/data/Transaction.d:
```diff
@@ -159,6 +136,29 @@ public struct Transaction
+
+    /// A `Coinbase` transaction is one that has one or more `Coinbase` outputs
+    /// However if all outputs are not `Coinbase` then it will fail validation
+    public bool isCoinbase () const
```
Ditto

hewison-chris replied:
Same

linked0 commented on source/agora/consensus/data/Transaction.d:
```diff
@@ -159,6 +136,29 @@ public struct Transaction
+    }
+
+    /// A `Payment` transaction is one that has outputs of type `Payment`
+    public bool isPayment () const
```
Ditto

hewison-chris replied:
Same

linked0 commented on source/agora/test/EnrollDifferentUTXO.d:
```diff
@@ -112,7 +113,7 @@ unittest
 
     // 8 utxos for freezing, 24 utxos for creating a block later
-    txs ~= spendable[4].split(WK.Keys.NODE2.address.repeat(8)).sign();
+    txs ~= spendable[4].split(genesis_validator_keys[0].address.repeat(8)).sign();
```
Can you explain why this should be changed again? I can not remember the reason.

hewison-chris replied:
I will extract this change into a separate commit with an explanation of the change. But it is basically because the test is using first node which is first enrollment in Genesis. NODE2 is not always the first node as the enrollments are ordered by utxo.

Geod24 reviewed (commented):
Forgot to post my comment...

linked0 reviewed (commented):
Good work. Some suggestions.

hewison-chris commented on source/agora/utils/TxBuilder.d:
```diff
@@ -257,7 +259,7 @@ public struct TxBuilder
     ***************************************************************************/
 
-    public Transaction sign (in TxType type = TxType.Payment, const(ubyte)[] data = [],
+    public Transaction sign (in OutputType outputs_type = OutputType.Payment, const(ubyte)[] data = [],
```
@Geod24  I went with this for now to make it clearer how `OutputType` is being used.

hewison-chris committed:
Unit test: Add info to assert failure

When the channel state is not matching one of the expected values in one
of the `test.Flash` unit tests it is useful to print the `ChannelState`.

hewison-chris commented:
Green again.

Geod24 commented on source/agora/consensus/data/Transaction.d:
```diff
@@ -260,10 +253,28 @@ public struct Output
             return this.lock < rhs.lock ? -1 : 1;
         return this.value.opCmp(rhs.value);
     }
+
+    /***************************************************************************
+
+        Transactions Serialization
+
+        Params:
+            dg = Serialize function
+
+    ***************************************************************************/
+
+    public void serialize (scope SerializeDg dg) const @trusted
+    {
+        serializePart(this.value, dg);
+        serializePart(this.lock, dg);
+        serializePart(this.type, dg);
+    }
```
Why ?

hewison-chris replied:
Removing this and re-ordering the fields in the Transaction. Also changing enum from `ubyte` to `unit`.

Geod24 commented on source/agora/consensus/data/Transaction.d:
```diff
@@ -206,22 +194,27 @@ public struct Output
 
     /// Ctor
-    public this (Amount value, inout(Lock) lock) inout pure nothrow @trusted
+    public this (Amount value, inout(Lock) lock, inout(OutputType) type = OutputType.Payment) inout pure nothrow @trusted
```
`in OutputType`

hewison-chris replied:
Can Amount also be `in` ?

Geod24 commented on source/agora/consensus/data/Transaction.d:
```diff
@@ -206,22 +194,27 @@ public struct Output
     public ulong sizeInBytes () const nothrow pure @safe @nogc
     {
-        return this.value.sizeof + this.lock.sizeInBytes();
+        return ubyte.sizeof + this.value.sizeof + this.lock.sizeInBytes();
```
`OutputType.sizeof + ...`

hewison-chris replied:
Updated

Geod24 replied:
Did you push ?

hewison-chris replied:
Have now

hewison-chris replied:
But still fixing up the tests.
